### PR TITLE
Add `Loop.replace`

### LINKED ISSUE #10309: Add `Loop.replace` to improve the customization UX
## 🚀 Feature

Implement a Loop method that allows easily replacing a Loop with not much customization.

### Motivation

There are 2 users of loops customization:
- (A): User that creates their own loop
- (B): User subclasses one of our loops but does not care about init args

Replacing the epoch loop is verbose for (B):

```python
trainer = Trainer(...)
# instantiate with existing values
epoch_loop = MyTrainingEpochLoop(trainer.fit_loop.epoch_loop.min_steps, trainer.fit_loop.epoch_loop.max_steps)
# connect children loops
epoch_loop.connect(trainer.fit_loop.epoch_loop.batch_loop, trainer.fit_loop.epoch_loop.val_loop)
# connect to parent loop
trainer.fit_loop.connect(epoch_loop)
```

This is tedious when the user made a simple modification, for example:

```python
class MyTrainingEpochLoop(TrainingEpochLoop):
    def _should_check_val_fx(self, *args, **kwargs):
        my_condition = self.trainer.optimizers[0].param_groups[0]["lr"] < 1e-3
        return my_condition and super()._should_check_val_fx(*args, **kwargs)
```

This proposal tries to make it easier for (B), but adds no benefits for (A)

### Pitch

Support:
```python
trainer.fit_loop.replace(MyTrainingEpochLoop)
```

where we find which `self` Loop attribute corresponds to this class, instantiates it with the correct arguments, connects the new loop to its subloops, and connects the new loop to the instance `replace` was called for. 

### Alternatives

Not do it, as it holds assumptions that will not work for any Loop implementation.

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @carmocca @justusschock @ananthsub @ninginthecloud

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes #10309

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @carmocca @justusschock @ananthsub @ninginthecloud

### CODE REVIEW

File: tests/loops/test_loops.py [hunk 1]
```diff
@@ -103,6 +103,36 @@ def test_connect_subloops(tmpdir):
...
+        trainer.fit_loop.replace(TestLoop)
+
+    with pytest.raises(MisconfigurationException, match="Did not find.*same parent class as `PredictionEpochLoop`"):
+        trainer.fit_loop.replace(PredictionEpochLoop)
+
+    class TestLoop(TrainingEpochLoop):
+        ...
+
+    old_loop = trainer.fit_loop.epoch_loop
+    new_loop = trainer.fit_loop.replace(TestLoop)
```
Comment (carmocca) [inline tests/loops/test_loops.py:125]: Another API option would be to have

```python
trainer.fit_loop.replace({"epoch_loop": TestLoop})
```

with the advantages:
- No need to look up the attribute internally
- Can replace multiple loops at the same time.
- Adds support for loops that have subloops of the same type

And disadvantage:
- Not as cool :P

File: pytorch_lightning/loops/base.py [hunk 2]
```diff
@@ -99,6 +99,51 @@ def connect(self, **kwargs: "Loop") -> None:
...
+                # compare the signatures
+                old_parameters = inspect.signature(old_loop.__class__.__init__).parameters
+                current_parameters = inspect.signature(type_or_object.__init__).parameters
+                if old_parameters != current_parameters:
+                    raise MisconfigurationException(
+                        f"`{self.__class__.__name__}.replace({type_or_object.__name__})` can only be used if the"
+                        f" `__init__` signatures match but `{old_loop.__class__.__name__}` does not."
+                    )
+                # instantiate the loop
+                kwargs = {p: getattr(old_loop, p) for p in old_parameters if p != "self"}
```
Comment (tchaton) [inline pytorch_lightning/loops/base.py:132]: Can we add a test for all our Loops to make sure the arguments are saved as attributes ?

File: pytorch_lightning/loops/base.py [hunk 3]
```diff
@@ -99,6 +99,51 @@ def connect(self, **kwargs: "Loop") -> None:
...
+                loop = type_or_object(**kwargs)
+            else:
+                loop = type_or_object
+
+            # connect sub-loops
+            kwargs = {n: l for n, l in old_loop.__dict__.items() if isinstance(l, Loop)}
```
Comment (tchaton) [inline pytorch_lightning/loops/base.py:138]: You are assuming the users didn't override the sub-loop if they provide an object. Should we filter them out?

Comment (carmocca) [reply pytorch_lightning/loops/base.py:138]: Mmm good question. I don't think we can know this reliably when an object is passed.

One thing we can do is that if an object is passed and `__init__` is overridden, we raise an exception saying it's not supported

File: pytorch_lightning/loops/base.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/loops/base.py:132]: Done

File: pytorch_lightning/loops/base.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/loops/base.py]: Another option is removing support for passing an instance, which completely avoids the problem. Users would need `connect` in that case

cc @justusschock @awaelchli

File: pytorch_lightning/loops/base.py [hunk 4]
```diff
@@ -99,6 +99,53 @@ def connect(self, **kwargs: "Loop") -> None:
...
 
+    def replace(self, **loops: Union["Loop", Type["Loop"]]) -> None:
+        """Optionally replace one or multiple of this loop's sub-loops.
+
+        This methods takes care of instantiating the class (if necessary) with all existing arguments, connecting all
+        sub-loops of the old loop to the new instance, setting the ``Trainer`` reference, and conneting the new loop to
+        the parent.
+
+        Args:
+            **loops: A ``Loop`` subclass or instance. The name used should match the loop attribute name you want to
```
Comment (awaelchli) [inline pytorch_lightning/loops/base.py:110]: here it is meant the name of the argument

File: pytorch_lightning/loops/base.py [hunk 5]
```diff
@@ -99,6 +99,53 @@ def connect(self, **kwargs: "Loop") -> None:
         Linked loops should form a tree.
         """
 
+    def replace(self, **loops: Union["Loop", Type["Loop"]]) -> None:
+        """Optionally replace one or multiple of this loop's sub-loops.
```
Comment (awaelchli) [inline pytorch_lightning/loops/base.py:103]: should we document that it calls connect()?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/loops/base.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/loops/base.py:103]: It's already mentioned below

File: pytorch_lightning/loops/base.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/loops/base.py]: yes, I mean to suggest making it clear the name mentioned here refers to the name of the argument. 

LGTM

File: pytorch_lightning/loops/base.py [hunk 6]
```diff
@@ -99,6 +99,62 @@ def connect(self, **kwargs: "Loop") -> None:
...
+
+        This methods takes care of instantiating the class (if necessary) with all existing arguments, connecting all
+        sub-loops of the old loop to the new instance, setting the ``Trainer`` reference, and ``connect``ing the new
```
Comment (carmocca) [inline pytorch_lightning/loops/base.py:106]:
```suggestion
        sub-loops of the old loop to the new instance, setting the ``Trainer`` reference, and ``connect``\ing the new
```

File: pytorch_lightning/loops/base.py [hunk 7]
```diff
@@ -99,6 +99,51 @@ def connect(self, **kwargs: "Loop") -> None:
         Linked loops should form a tree.
         """
 
+    def replace(self, **loops: Union["Loop", Type["Loop"]]) -> None:
+        """Optionally replace one or multiple of this loop's sub-loops.
+
+        This methods takes care of instantiating the class (if necessary) with all existing arguments, connecting all
+        sub-loops of the old loop to the new instance, setting the ``Trainer`` reference, and conneting the new loop to
```
Comment (rohitgr7) [inline pytorch_lightning/loops/base.py:106]:
```suggestion
        sub-loops of the old loop to the new instance, setting the ``Trainer`` reference, and connecting the new loop to
```

File: pytorch_lightning/loops/base.py [hunk 8]
```diff
@@ -99,6 +99,51 @@ def connect(self, **kwargs: "Loop") -> None:
...
+        the parent.
+
+        Args:
+            **loops: A ``Loop`` subclass or instance. The name used should match the loop attribute name you want to
```
Comment (rohitgr7) [inline pytorch_lightning/loops/base.py:110]: `A` loop?

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/loops/base.py (hunk 8 above)
Comment (carmocca) [reply pytorch_lightning/loops/base.py:110]: What do you suggest exactly? Note these are `**kwargs` so you pass them individually.

Comment (rohitgr7) [reply pytorch_lightning/loops/base.py:110]: A sequence of ``Loop``(s), maybe?

**Review by justusschock — APPROVED** [review]
